csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -0,0 +1,95 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <debug.h>
+#include <ir/utils.h>
+#include <options.h>
+#include <preseg_passes/remove_bcast_squeeze.h>
+#include <root_domain_map.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+inline bool isFusionOutput(Fusion* fusion, Val* val) {
+  const auto& outputs = fusion->outputs();
+  return std::find(outputs.begin(), outputs.end(), val) != outputs.end();
+}
+
+// Remove broadcast-squeeze and squeeze-broadcast patterns
+// TODO: still remove when have intermediate ops between broadcast and squeeze
+void removeBcastSqueeze(Fusion* fusion) {
+  // step-1: find and remove broadcast + squeeze pattern
+  // before: Y = broadcast(X); Z = squeeze(Y);  M = someOp(Z)
+  // after : M = someOp(X)
+  // conditions: (1) broadcast and squeeze have the same dim flags
+  //             (2) Y is only consumed by Z (has only one consumer)
+  // special case: if Z is a fusion output, replace the output with X
+  {
+    std::unordered_map<Val*, Val*> replacement_map;
+    for (auto expr : fusion->exprs()) {
+      if (auto sop = dynamic_cast<SqueezeOp*>(expr)) {
+        if (auto bcast = dynamic_cast<BroadcastOp*>(sop->in()->definition())) {
+          if (bcast->getBroadcastDimFlags() == sop->getSqueezeDimFlags() &&
+              bcast->out()->uses().size() == 1) {
+            if (isFusionOutput(fusion, sop->out())) {
+              fusion->replaceOutput(sop->out(), bcast->in());
+            }
+            replacement_map.insert({sop->out(), bcast->in()});
+          }
+        }
+      }
+    }
+    // Replace if there is any match
+    if (!replacement_map.empty()) {
+      ir_utils::replaceValue(fusion, replacement_map);
```

Comment:
Have you considered ir_utils::replaceValInAllExprInputsAndFusionOutputs?

Reply:
changed to use `ir_utils::replaceValInAllExprInputsAndFusionOutputs` and merged two loops into one.

Reply:
I need to roll back to use `ir_utils::replaceValue(fusion, replacement_map);` to fix the failed test case.
I added a cpp unit to reproduce the failed python test, `BcastSqueezeBcastSqueeze`.
In the following fusion, 
```
  auto tv2 = broadcast(tv1, {false, false, true});
  auto tv3 = squeeze(tv2, std::vector<bool>{false, false, true});
  auto tv4 = broadcast(tv3, {false, false, true});
  auto tv5 = squeeze(tv4, std::vector<bool>{false, false, true});
  auto tv6 = add(tv3, tv5);
```
(1) first step: replace `tv3` with `tv1` 
(2) then, replace `tv5` with `tv1`, 
(3) then, the segment error happens when dynamic_cast expr `auto tv6 = add(tv3, tv5)`
```
Caught signal 11 (Segmentation fault: address not mapped to object at address 0xfffffffffffffff0)
==== backtrace (tid: 488215) ====
 0 0x0000000000042520 __sigaction()  ???:0
 1 0x00000000000acc9d __dynamic_cast()  ???:0
 2 0x00000000006a6e82 nvfuser::preseg_passes::RemoveBcastSqueeze::runPass()  :0
```
It looks like `replaceValInExprInputs` invalidates provided Expr.
```
//! Replaces reference Val with substitute in all Expr inputs and attributes.
//! Warning: Invalidates provided Expr.
//! Warning: Removes connection of reference through provided Expr.
//! Warning: Creates new Expr defining substitute.
NVF_API Expr* replaceValInExprInputs(
    Expr* expr,
    Val* reference,
    Val* substitute);
```

Reply:
Thanks -- that's something I overlooked. Can you try this? https://github.com/NVIDIA/Fuser/blob/25044927da7b5f4a0cf3fb16ee716dc0403b46b5/csrc/multidevice/utils.cpp#L279-L283

Reply:
smart solution!